crates/rpc/rpc/src/eth/api/transactions.rs

```diff
@@ -5,32 +5,64 @@ use crate::{
 };
 use async_trait::async_trait;
 use reth_primitives::{
-    BlockId, Bytes, FromRecoveredTransaction, IntoRecoveredTransaction, TransactionSigned,
-    TransactionSignedEcRecovered, H256, U256,
+    BlockId, BlockNumberOrTag, Bytes, FromRecoveredTransaction, IntoRecoveredTransaction,
+    TransactionSigned, TransactionSignedEcRecovered, H256, U256,
 };
-use reth_provider::{BlockProvider, EvmEnvProvider, StateProviderFactory, TransactionsProvider};
+use reth_provider::{BlockProvider, EvmEnvProvider, StateProviderFactory};
 use reth_rlp::Decodable;
 use reth_rpc_types::{Index, Transaction, TransactionRequest};
 use reth_transaction_pool::{TransactionOrigin, TransactionPool};
+use revm::primitives::{BlockEnv, CfgEnv};
 
 /// Commonly used transaction related functions for the [EthApi] type in the `eth_` namespace
 #[async_trait::async_trait]
 pub trait EthTransactions: Send + Sync {
+    /// Returns the revm evm env for the requested [BlockId]
+    ///
+    /// If the [BlockId] this will return the [BlockId::Hash] of the block the env was configured
+    /// for.
+    async fn evm_env_at(&self, at: BlockId) -> EthResult<(CfgEnv, BlockEnv, BlockId)>;
+
     /// Returns the transaction by hash.
     ///
     /// Checks the pool and state.
     ///
     /// Returns `Ok(None)` if no matching transaction was found.
     async fn transaction_by_hash(&self, hash: H256) -> EthResult<Option<TransactionSource>>;
+
+    /// Returns the transaction by including its corresponding [BlockId]
+    async fn transaction_by_hash_at(
+        &self,
+        hash: H256,
+    ) -> EthResult<Option<(TransactionSource, BlockId)>>;
 }
 
 #[async_trait]
 impl<Client, Pool, Network> EthTransactions for EthApi<Client, Pool, Network>
 where
     Pool: TransactionPool + Clone + 'static,
-    Client: TransactionsProvider + 'static,
+    Client: BlockProvider + StateProviderFactory + EvmEnvProvider + 'static,
     Network: Send + Sync + 'static,
 {
+    async fn evm_env_at(&self, at: BlockId) -> EthResult<(CfgEnv, BlockEnv, BlockId)> {
+        // TODO handle Pending state's env
+        match at {
+            BlockId::Number(BlockNumberOrTag::Pending) => {
+                // This should perhaps use the latest env settings and update block specific
+                // settings like basefee/number
+                unimplemented!("support pending state env")
+            }
+            hash_or_num => {
+                let block_hash = self
+                    .client()
+                    .block_hash_for_id(hash_or_num)?
+                    .ok_or_else(|| EthApiError::UnknownBlockNumber)?;
+                let (cfg, env) = self.cache().get_evm_env(block_hash).await?;
+                Ok((cfg, env, block_hash.into()))
```

Comment:
is block hash lookup needed? why not just pass `hash_or_num` to the fn?

Reply:
Because it might be a non existent block perhaps?